src/truncated/lognormal.jl

```diff
@@ -0,0 +1,50 @@
+# Moments of the truncated log-normal can be computed directly from the moment generating
+# function of the truncated normal:
+#   Let Y ~ LogNormal(μ, σ) truncated to (a, b). Then log(Y) ~ Normal(μ, σ) truncated
+#   to (log(a), log(b)), and E[Y^n] = E[(e^log(Y))^n] = E[e^(nlog(Y))] = mgf(log(Y), n).
+
+# Given `truncate(LogNormal(μ, σ), a, b)`, return `truncate(Normal(μ, σ), log(a), log(b))`
+function _truncnorm(d::Truncated{<:LogNormal})
+    μ, σ = params(d.untruncated)
+    T = float(partype(d))
+    a = d.lower === nothing || d.lower <= 0 ? nothing : log(T(d.lower))
+    b = d.upper === nothing || isinf(d.upper) ? nothing : log(T(d.upper))
```

Comment:
Tests on Julia 1.3 pass locally when changing this to
```suggestion
    a = d.lower === nothing ? nothing : log(T(max(d.lower, 0)))
    b = d.upper === nothing ? nothing : log(T(d.upper))
```

Reply:
That isn't functionally equivalent though; IIRC, I was relying on `a` and/or `b` being `nothing` in those cases so that `truncated` would handle it a particular way.

Reply:
But arguably this optimization (`d.lower === nothing` or `d.upper === nothing`) should already have happened, either by a user or internally, when constructing the `d = truncated(LogNormal(...))`? Maybe one shouldn't expect that unoptimized inputs lead to optimized algorithms. AFAICT the optimization of `truncated(Normal(...), ...)` is also only exploited in the case where this returns a `Normal` (`a = b = nothing`); the code for `Truncated{<:Normal}` does not seem to use the fact that a bound might be `nothing`. In the `Normal` case, arguably the `LogNormal` shouldn't be truncated in the first place.

Reply:
I'm currently sick and can only barely comprehend that message but if you'd prefer to go with your suggested change then feel free to apply it, I trust your judgement

Reply:
Did we have this conversation a few months ago? My brain similarly can't comprehend whether this is the same discussion: https://github.com/JuliaStats/Distributions.jl/pull/1874#discussion_r1659372206

I should probably just log off and go sleep